universal-hash/src/lib.rs

```diff
@@ -18,167 +18,153 @@
 //! [Universal Hash Functions]: https://en.wikipedia.org/wiki/Universal_hashing
 
 #![no_std]
-#![forbid(unsafe_code)]
 #![doc(
     html_logo_url = "https://raw.githubusercontent.com/RustCrypto/media/8f1a9894/logo.svg",
     html_favicon_url = "https://raw.githubusercontent.com/RustCrypto/media/8f1a9894/logo.svg",
-    html_root_url = "https://docs.rs/universal-hash/0.4.1"
+    html_root_url = "https://docs.rs/universal-hash/0.5.0"
 )]
+#![cfg_attr(docsrs, feature(doc_cfg))]
 #![warn(missing_docs, rust_2018_idioms)]
 
 #[cfg(feature = "std")]
 extern crate std;
 
-pub use generic_array::{self, typenum::consts};
+pub use crypto_common::{
+    self, generic_array,
+    typenum::{self, consts},
+    Block, Key, KeyInit, ParBlocks,
+};
 
-use generic_array::typenum::Unsigned;
+use core::slice;
+use crypto_common::{BlockSizeUser, ParBlocksSizeUser};
 use generic_array::{ArrayLength, GenericArray};
-use subtle::{Choice, ConstantTimeEq};
-
-/// Keys to a [`UniversalHash`].
-pub type Key<U> = GenericArray<u8, <U as NewUniversalHash>::KeySize>;
-
-/// Blocks are inputs to a [`UniversalHash`].
-pub type Block<U> = GenericArray<u8, <U as UniversalHash>::BlockSize>;
-
-/// Instantiate a [`UniversalHash`] algorithm.
-pub trait NewUniversalHash: Sized {
-    /// Size of the key for the universal hash function.
-    type KeySize: ArrayLength<u8>;
+use subtle::ConstantTimeEq;
+use typenum::Unsigned;
+
+/// Trait implemented by UHF backends.
+pub trait UhfBackend: ParBlocksSizeUser {
+    /// Process single block.
+    fn proc_block(&mut self, block: &Block<Self>);
```

Comment:
Per https://github.com/RustCrypto/universal-hashes/pull/153#discussion_r857124619, we are currently lacking a way to inform the consumer of the `UhfBackend` that the backend is "unaligned" and cannot consume `ParBlocks` efficiently (with impls either panicking, or always falling back to a simple loop over `UhfBackend::proc_block` without any parallelism). One way to avoid this would be for the backend to expose the misalignment:
```suggestion
    fn proc_block(&mut self, block: &Block<Self>);

    /// Returns the number of blocks that should be passed to `Self::proc_block` before
    /// `Self::proc_par_blocks` can be used efficiently. This is always less than
    /// `Self::ParBlocksSize`.
    fn blocks_needed_to_align(&self) -> usize;
```